File: src/main/java/net/atos/zac/admin/ReferenceTableService.java
/*
 * SPDX-FileCopyrightText: 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.admin;

import static jakarta.transaction.Transactional.TxType.SUPPORTS;

import java.util.List;
import java.util.Optional;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.persistence.EntityManager;
import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.CriteriaQuery;
import jakarta.persistence.criteria.Root;
import jakarta.transaction.Transactional;

import net.atos.zac.admin.model.ReferenceTable;

@ApplicationScoped
@Transactional(SUPPORTS)
public class ReferenceTableService {
    private EntityManager entityManager;

    @Inject
    public ReferenceTableService(final EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    /**
     * Default no-arg constructor, required by Weld.
     */
    public ReferenceTableService() {
    }

    public List<ReferenceTable> listReferenceTables() {
        final CriteriaBuilder builder = entityManager.getCriteriaBuilder();
        final CriteriaQuery<ReferenceTable> query = builder.createQuery(ReferenceTable.class);
        final Root<ReferenceTable> root = query.from(ReferenceTable.class);
        query.orderBy(builder.asc(root.get("naam")));
        query.select(root);
        return entityManager.createQuery(query).getResultList();
    }

    public ReferenceTable readReferenceTable(final long id) {
        final ReferenceTable referenceTable = entityManager.find(ReferenceTable.class, id);
        if (referenceTable != null) {
            return referenceTable;
        } else {
            throw new RuntimeException("Reference table with id '%d' not found".formatted(id));
        }
    }

    public ReferenceTable readReferenceTable(final String code) {
        return findReferenceTable(code)
                .orElseThrow(
                        () -> new RuntimeException("Reference table with code '%s' not found".formatted(code))
                );
    }

    public Optional<ReferenceTable> findReferenceTable(final String code) {
        final CriteriaBuilder builder = entityManager.getCriteriaBuilder();
        final CriteriaQuery<ReferenceTable> query = builder.createQuery(ReferenceTable.class);
        final Root<ReferenceTable> root = query.from(ReferenceTable.class);
        query.select(root).where(builder.equal(root.get("code"), code));
        final List<ReferenceTable> resultList = entityManager.createQuery(query).getResultList();
        return resultList.isEmpty() ? Optional.empty() : Optional.of(resultList.getFirst());
    }
}

File: src/main/kotlin/net/atos/zac/admin/ReferenceTableService.kt
/*
 * SPDX-FileCopyrightText: 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.admin

import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.persistence.EntityManager
import jakarta.transaction.Transactional
import net.atos.zac.admin.exception.ReferenceTableNotFoundException
import net.atos.zac.admin.model.ReferenceTable
import nl.info.zac.util.AllOpen
import nl.info.zac.util.NoArgConstructor

@ApplicationScoped
@NoArgConstructor
@AllOpen
@Transactional(Transactional.TxType.SUPPORTS)
class ReferenceTableService @Inject constructor(
    val entityManager: EntityManager
) {
    /**
     * Finds a reference table by its code, where code is treated as uppercase only.
     *
     * @return the reference table or null if none could be found
     */
    @Suppress("NestedBlockDepth")
    fun findReferenceTable(code: String): ReferenceTable? =
        entityManager.criteriaBuilder.let { criteriaBuilder ->
            criteriaBuilder.createQuery(ReferenceTable::class.java).let { query ->
                query.from(ReferenceTable::class.java).let { root ->
                    criteriaBuilder.equal(root.get<Any>("code"), code.uppercase()).let { predicate ->
                        query.select(root).where(predicate)
                    }
                }
                entityManager.createQuery(query).resultList
            }
        }.firstOrNull()

    fun listReferenceTableValuesSorted(referenceTable: ReferenceTable) =
        referenceTable.values
            .sortedBy { it.sortOrder }
            .toList()

    fun listReferenceTables(): List<ReferenceTable> =
        entityManager.criteriaBuilder.let { criteriaBuilder ->
            criteriaBuilder.createQuery(ReferenceTable::class.java).let { query ->
                query.from(ReferenceTable::class.java).let {
                    query.orderBy(criteriaBuilder.asc(it.get<Any>("name")))
                    query.select(it)
                }
                entityManager.createQuery(query).resultList
            }
        }

    fun readReferenceTable(id: Long): ReferenceTable =
        entityManager.find(ReferenceTable::class.java, id) ?: run {
            throw ReferenceTableNotFoundException(id)
        }

    fun readReferenceTable(code: String): ReferenceTable =
        findReferenceTable(code) ?: run {
            throw ReferenceTableNotFoundException("No reference table found with code '$code'")
        }
}
